### PR TITLE
refactor: clean up and introduce thread workers

### PR DESCRIPTION
**Pull request by smusali**
**ci: cleaning on automation**

Semver: patch

---
**refactor: reduce code complexity**

Semver: patch

---
**feat(threadWorkerPools): introduce extra threads**

In order to provide non-blocking I/O, introducing ThreadWorkerPools for multiple parts and imitating the non-blocking I/O implementation from [our Ruby library](github.com/logdna/ruby).

Semver: minor

---
**test(logger): modify testing for logger**

Semver: patch

---
**feat(meta): enable adding custom meta fields**

This commit introduces a new option called `custom_fields` similar to `tags` and also new default `meta` fields: `lineno`, `name`, `pathname`, `args`, etc.

Semver: minor

---
**docs: update the README as requested**

Semver: patch

### CODE REVIEW

File: logdna/logdna.py [hunk 1]
```diff
@@ -1,190 +1,264 @@
...
+            self.start_flusher()
 
-    # do not call without acquiring the lock
-    def send_request(self):
+    def try_request(self):
         self.buf.extend(self.secondary)
         self.secondary = []
         data = {'e': 'ls', 'ls': self.buf}
+        retries = 0
+        while True:
```
Comment (teeohhem) [inline logdna/logdna.py:169]: Could write this as:
```
while retries < self.max_retries_attempts:
```

File: logdna/utils.py [hunk 2]
```diff
@@ -10,7 +10,15 @@ def is_jsonable(obj):
...
 
-def sanitize_meta(meta):
+def sanitize_meta(meta, index_meta=False):
+    if not index_meta:
+        if is_jsonable(meta):
+            return json.dumps(meta)
+
+        return {
+            '__errors': 'Meta cannot be serialized into JSON-formatted string'
+        }
```
Comment (teeohhem) [inline logdna/utils.py:20]: You may want to throw a try/except around `json.dumps`. If that fails for any reason, then it'll raise an exception that won't be handled here

**Comment by teeohhem** [comment]
Overall, nice work! This has a lot of great changes. I've added a few comments to improve some things. Also, we'll need a ticket to restore some of the integration test functionality. The unit tests only cover so much.

File: logdna/utils.py (hunk 2 above)
Comment (smusali) [reply logdna/utils.py:20]: That's been tested within `is_jsonable` function; so, no need to add extra `try/except`

File: logdna/logdna.py [hunk 3]
```diff
@@ -1,190 +1,265 @@
...
+        self.include_fields = options.get('include_fields', [])
+        if isinstance(self.include_fields, str):
+            self.include_fields = [
+                include_field.strip()
+                for include_field in self.include_fields.split(',')
+            ]
+        elif not isinstance(self.include_fields, list):
+            self.include_fields = []
+        self.include_fields += defaults['META_FIELDS']
+
```
[the remark this answers is not in the document]
Comment (smusali) [reply logdna/logdna.py:37]: Done!

File: logdna/logdna.py (hunk 1 above)
Comment (smusali) [reply logdna/logdna.py]: Done!

File: tests/test_logger.py [hunk 4]
```diff
@@ -1,112 +1,130 @@
...
+        handler = LogDNAHandler(LOGDNA_API_KEY, sample_options)
+        self.assertIsNone(handler.flusher)
+        handler.start_flusher()
+        self.assertIsNotNone(handler.flusher)
+        handler.close_flusher()
+        self.assertIsNone(handler.flusher)
+
+    def emit_test(self):
+        handler = LogDNAHandler(LOGDNA_API_KEY, sample_options)
+        handler.buffer_log = unittest.mock.Mock()
```
Comment (esatterwhite) [inline tests/test_logger.py:95]: This is meat of functionality, we shouldn't mock this. This leaves the coverage of the logger itself at ~50%

Comment (smusali) [reply tests/test_logger.py:95]: That's correct! `buffer_log` is the core of the library but, @esatterwhite, in this case, what would your suggestion be in terms of testing `emit`? CC: @teeohhem

Comment (teeohhem) [reply tests/test_logger.py:95]: Personally, if we're going to forgo the integration tests for now, I'd look into implementing a mock `ThreadPoolExecutor` that returns synchronously.

e.g.

```
class MockThreadPoolExecutor():
    def __init__(self, **kwargs):
        pass

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        pass

    def submit(self, fn, *args, **kwargs):
        # execute functions in series without creating threads
        # for easier unit testing
        result = fn(*args, **kwargs)
        return result

    def shutdown(self, wait=True):
        pass
```

Then in the unit test you can do something like this:
```
with patch('logdna.ThreadPoolExecutor', wraps= MockThreadPoolExecutor()) as mock:
    #write tests against function and assert buffer contains x

Comment (smusali) [reply tests/test_logger.py:115]: With the help of `MockThreadPoolExecutor`, thanks to @teeohhem, it's now 75+%

**Review by teeohhem — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -67,10 +67,9 @@ _**Required**_
 * [LogDNA Ingestion Key](https://app.logdna.com/manage/profile)
 
 _**Optional**_
-* Hostname - ([string][]) - max length 32 chars
+* Hostname - ([string][])
 * MAC Address - ([string][])
```
Comment (tanberry) [inline README.md:71]: @Samir, I have some general comments, on sections above (but of course GH doesn't let me comment on "old" sections):

1. line # 17: In the header let's use **fileConfig** as one word, like it is in the body content.
2. Generally, across the doc, let's use the angle brackets format for any variables, so for example <my sample log line> or <my custom app>.
3. line # 136 needs a period at end of sentence.

File: README.md [hunk 6]
```diff
@@ -156,7 +155,6 @@ The [LogDNA API Key](https://app.logdna.com/manage/profile) associated with your
 * Type: [string][]
 * Default: `''`
 * Values: `YourCustomApp`
-* Max Length: `32`
```
Comment (tanberry) [inline README.md:159]: If `YourCustomApp` is a variable, let's use `<your custom app>`

File: README.md [hunk 7]
```diff
@@ -203,14 +200,12 @@ If this option is set to True then meta objects are parsed and searchable up to
 
 *WARNING* If this option is True, your metadata objects MUST have consistent types across all log messages or the metadata object may not be parsed properly.
```
Comment (tanberry) [inline README.md:202]: Use "might" instead o f"may." (The reason is that "may" implies permission and "might"implies chance or possibility.)

**Review by tanberry — APPROVED** [review]
Just a few comments, take what you like and leave the rest. ;-)

File: README.md [hunk 8]
```diff
@@ -156,7 +155,6 @@ The [LogDNA API Key](https://app.logdna.com/manage/profile) associated with your
 * Type: [string][]
 * Default: `''`
 * Values: `YourCustomApp`
-* Max Length: `32`
 
 The default app passed along with every log sent through this instance.
```
[the remark this answers is not in the document]
Comment (smusali) [reply README.md:160]: Yes!

File: README.md (hunk 5 above)
Comment (smusali) [reply README.md:71]: Done, @tanberry!

File: README.md (hunk 6 above)
Comment (smusali) [reply README.md:159]: Done across all, @tanberry!

File: README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (smusali) [reply README.md:160]: Done, @tanberry!

File: README.md (hunk 7 above)
Comment (smusali) [reply README.md:202]: Done, @tanberry!

**Comment by smusali** [comment]
@tanberry, updated as you requested!

**Review by teeohhem — APPROVED** [review]

**Review by tanberry — APPROVED** [review]

**Comment by smusali** [comment]
These changes are for resolving and/or improving #61, #59, #58, and #56.

**Pull request merged**